Source/Graph/Graph.cs

```diff
@@ -623,45 +623,66 @@ public List<Node> SuccessorsAsList(Node n)
+    /// (using the property: A vertex's dominator set is unaffected by vertices that come later).
+    /// The method does not check the graph for the DAG property. That risk is on the caller.
+    /// </summary>
+    public Dictionary<Node, HashSet<Node>> AcyclicDominators()
     {
-      var topoSorted = TopologicalSort().ToList();
-      var dominators = new Dictionary<Node, HashSet<Node>>();
-      topoSorted.ForEach(u => dominators[u] = topoSorted.ToHashSet());
-      foreach (var node in topoSorted)
+      var dominatorsPerNode = new Dictionary<Node, HashSet<Node>>();
+      foreach (var node in TopologicalSort())
       {
-        var s = new HashSet<Node>();
-        var predecessors = Predecessors(node).ToList();
-        if (predecessors.Count != 0)
-        {
-          s.UnionWith(dominators[predecessors.First()]);
-          predecessors.ForEach(v => s.IntersectWith(dominators[v]));
+        var predecessors = Predecessors(node);
+        var dominatorsForNode = Intersection(predecessors.Select(p => dominatorsPerNode[p]));
+        dominatorsForNode.Add(node);
+        dominatorsPerNode[node] = dominatorsForNode;
+      }
+      return dominatorsPerNode;
+    }
+
+    public static HashSet<T> Intersection<T>(IEnumerable<ISet<T>> sets) {
+      var first = true;
+      HashSet<T> result = null;
+      foreach (var set in sets) {
+        if (first) {
+          result = set.ToHashSet();
+          first = false;
+        } else {
+          result!.IntersectWith(set);
         }
-        s.Add(node);
-        dominators[node] = s;
       }
-      return dominators;
+
+      if (result == null) {
+        return new HashSet<T>();
+      }
+
+      return result;
     }
 
-    // Use this method only for DAGs because it uses DominatorsFast() for computing dominators
+    /// <summary>
+    /// Use this method only for DAGs because it uses DominatorsFast() for computing dominators
+    /// </summary>
     public Dictionary<Node, Node> ImmediateDominator()
     {
-      List<Node> topoSorted = TopologicalSort().ToList();
-      Dictionary<Node, HashSet<Node>> dominators = DominatorsFast();
+      var topoSorted = TopologicalSort().ToList();
+      var indexPerNode = new Dictionary<Node, int>();
+      for (int index = 0; index < topoSorted.Count; index++) {
+        indexPerNode[topoSorted[index]] = index;
+      }
+      var dominators = AcyclicDominators();
       var immediateDominator = new Dictionary<Node, Node>();
       foreach (var node in Nodes)
       {
         if (dominators[node].Count > 1)
         {
           dominators[node].Remove(node);
         }
-        immediateDominator[node] = topoSorted.ElementAt(dominators[node].Max(e => topoSorted.IndexOf(e)));
+        immediateDominator[node] = topoSorted.ElementAt(dominators[node].Max(e => indexPerNode[e]));
       }
-      immediateDominator[source] = source;
+
+      immediateDominator.Remove(source);
```

Comment:
I don't have the algorithmic context to understand these changes but I'll accept it as-it, hopefully it's well-tested in the CI.

Reply:
The behavior of the method is slightly changed: it no longer inserts the source node into the mapping, which corresponds better to the definition of immediate denominator which says "Every node, except the entry node, has an immediate dominator."